src/adapter/src/coord.rs

```diff
@@ -1734,16 +1810,30 @@ impl Coordinator {
 
         let time = write_frontier.clone().into_option().expect("checked above");
         let time = time.saturating_sub(lag);
-        let max_as_of = Antichain::from_elem(time);
+        let compaction_frontier = Antichain::from_elem(time);
+
+        // We must not select an `as_of` that is beyond any times that have not yet been written to
+        // downstream materialized views. If we would, we might skip times in the output of these
+        // materialized views, violating correctness. So our chosen `as_of` must be at most the
+        // meet of the `upper`s of all dependent materialized views.
+        let id_bundle = CollectionIdBundle {
+            storage_ids: dependent_matviews,
+            ..Default::default()
+        };
+        let max_as_of = self.least_valid_write(&id_bundle);
 
-        let as_of = min_as_of.join(&max_as_of);
+        let mut as_of = min_as_of.clone();
+        as_of.join_assign(&compaction_frontier);
+        as_of.meet_assign(&max_as_of);
```

Comment:
Should we add a warning if `max_as_of < compaction_frontier`. I think that we are assuming here that this never happens.

Reply:
We are not! It does happen, and in fact that is what caused the bug this PR is intended to fix.

`compaction_frontier` here is the compaction window applied to the `write_frontier` (which in turn is the minimum of all input write frontiers). Whereas `max_as_of` is the minimum write frontier of dependent materialized views. And there is nothing that guarantees that the latter is beyond the former.

 Perhaps the naming is confusing? `compaction_frontier` implies that input data before this frontier has been compacted, which isn't necessarily true. I'll rename it to `max_compaction_frontier`.